client: update expiration check to match spec

joshuagl opened the pull request:
Fixes #1231

**Description of the changes being introduced by the pull request**:

The specification, as of 1.0.16, describes an update expiration check as:

> The expiration timestamp in the trusted $ROLE metadata file MUST be
  higher than the fixed update expiration time.

Having done some research into how other security providers are comparing
expiration equivalents (i.e. OpenSSL x509 certificate checking code, and
GnuPG expiration checks), and how other TUF implementations are performing
the same check (rust-tuf, go-tuf), we came to a consensus that the correct
way to implement expiration comparisons is:

    expiration <= now

Where:
  expiration: is the metadata's expiration datetime
  now: is the current system time, or the fixed notion of time in the
       detailed client workflow (introduced in 1.0.16 of the spec)

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

joshuagl committed:
client: update expiration check to match spec

The specification, as of 1.0.16, describes an update expiration check as:

> The expiration timestamp in the trusted $ROLE metadata file MUST be
  higher than the fixed update expiration time.

Having done some research into how other security providers are comparing
expiration equivalents (i.e. OpenSSL x509 certificate checking code, and
GnuPG expiration checks), and how other TUF implementations are performing
the same check (rust-tuf, go-tuf), we came to a consensus that the correct
way to implement expiration comparisons is:

    expiration <= now

Where:
  expiration: is the metadata's expiration datetime
  now: is the current system time, or the fixed notion of time in the
       detailed client workflow (introduced in 1.0.16 of the spec)

Fixes #1231

Signed-off-by: Joshua Lock <jlock@vmware.com>

trishankatdatadog reviewed (commented):
LGTM, thanks, Joshua!

Could we quickly mock some tests, please, if that's not too much to ask?

Might also help to test we're handling UTC correctly, it was not obvious to me from reading the code...

joshuagl committed:
Update tests for client expiration check

Add a test to ensure that metadata expires at the expiration time, not
after it.
This tests the change to the updater introduced in 4bcd703

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
> Could we quickly mock some tests, please, if that's not too much to ask?

Done in follow-on commit fccd078

> Might also help to test we're handling UTC correctly, it was not obvious to me from reading the code...

I'm not sure the test I added is doing this, could you elaborate a bit more on what your concern is? Perhaps we can add a test to check.

trishankatdatadog commented:
> I'm not sure the test I added is doing this, could you elaborate a bit more on what your concern is? Perhaps we can add a test to check.

I think your tests cover `<`, `==`, and `>`, so we should be OK.

MVrachev reviewed (approved):
LGTM!
